### PR TITLE
redirect uri from route

### LINKED ISSUE #37: Add redirect_uri param to requests
I found out that this package has a bug: if multiple redirect URIs are configured on the app on Azure portal, the last one is used to redirect back.

`redirect_uri` param should be added to the azure url and the following requests. Adding that fixes this issue.

Should I go ahead and make a PR? Should we add the `redirect_uri` as a configurable param, or just refer to the "login/azurecallback" route (maybe transforming that into a named route)?

### PR DESCRIPTION
**Pull request by jimmypuckett**
This will Fix #37 

I am using a named route `azure.callback` when defining the route, so there is really nothing extra to configure.

### CODE REVIEW

**Comment by jimmypuckett** [comment]
@danieltjewett I got to thinking about this change, & it would be breaking for someone that already had the package installed without the named route, so I am not sure if you would want to put logic testing for the named route or tag a minor version change?

File: src/Azure.php [hunk 1]
```diff
@@ -53,6 +53,7 @@ public function handle($request, Closure $next)
                     'client_secret' => config('azure.client.secret'),
                     'refresh_token' => $refresh_token,
                     'resource' => config('azure.resource'),
+                    'redirect_uri' => route('azure.callback'),
```
Comment (danieltjewett) [inline src/Azure.php]: Make this an optional parameter.  Perhaps move `form_param` into its own hash above the try catch.  Then, check to see if the `route('azure.callback`)` exists.  If it does, add the key.

File: src/Azure.php [hunk 2]
```diff
@@ -98,7 +99,7 @@ protected function handleTesting(Request $request, Closure $next)
      */
     public function getAzureUrl()
     {
-        return $this->baseUrl . config('azure.tenant_id') . $this->route2 . "authorize?response_type=code&client_id=" . config('azure.client.id') . "&domain_hint=" . urlencode(config('azure.domain_hint')) . "&scope=" . urldecode(config('azure.scope'));
+        return $this->baseUrl . config('azure.tenant_id') . $this->route2 . "authorize?response_type=code&client_id=" . config('azure.client.id') . "&domain_hint=" . urlencode(config('azure.domain_hint')) . "&scope=" . urldecode(config('azure.scope'))  . '&redirect_uri=' . urlencode(route('azure.callback'));
```
Comment (danieltjewett) [inline src/Azure.php]: Let's create a string called `url` and then set it to what it was originally.  Then, check to see if the `azure.callback` route exists and if it does, append as suggested.

File: README.md [hunk 3]
```diff
@@ -8,8 +8,10 @@ Provides Azure Authentication Middleware for a Laravel App.  If you like this, c
 2. run `php artisan vendor:publish --provider="RootInc\LaravelAzureMiddleware\AzureServiceProvider"` to install config file to `config/azure.php`
 3. In our routes folder (most likely `web.php`), add
 ```php
-Route::get('/login/azure', '\RootInc\LaravelAzureMiddleware\Azure@azure');
-Route::get('/login/azurecallback', '\RootInc\LaravelAzureMiddleware\Azure@azurecallback');
+Route::get('/login/azure', '\RootInc\LaravelAzureMiddleware\Azure@azure')
```
Comment (danieltjewett) [inline README.md:11]: Let's add a step (3.a, 3.b?, or just a new int) that mentioned the possibility of naming routes.  If we name the route, we don't have to configure the callback on the portal.  If we don't, we need to make sure we add the URL in the portal.

File: README.md [hunk 4]
```diff
@@ -86,9 +88,12 @@ The above gives us a way to add/update users after a successful handshake.  `$p
 3. Our routes need to be updated to the following:
 
 ```php
-Route::get('/login/azure', '\App\Http\Middleware\AppAzure@azure');
-Route::get('/login/azurecallback', '\App\Http\Middleware\AppAzure@azurecallback');
-Route::get('/logout/azure', '\App\Http\Middleware\AppAzure@azurelogout');
+Route::get('/login/azure', '\App\Http\Middleware\AppAzure@azure')
+    ->name('azure.login');
+Route::get('/login/azurecallback', '\App\Http\Middleware\AppAzure@azurecallback')
```
Comment (danieltjewett) [inline README.md:95]: I would just add a note saying `if we are using named routes, be sure to add the named routes here appropriately.`

File: README.md [hunk 5]
```diff
@@ -202,7 +207,7 @@ class AppAzure extends Azure
     //we could overload this if we wanted too.
     public function getAzureUrl()
     {
-        return $this->baseUrl . config('azure.tenant_id') . $this->route2 . "authorize?response_type=code&client_id=" . config('azure.client.id') . "&domain_hint=" . urlencode(config('azure.domain_hint')) . "&scope=" . urldecode(config('azure.scope'));
+        return $this->baseUrl . config('azure.tenant_id') . $this->route2 . "authorize?response_type=code&client_id=" . config('azure.client.id') . "&domain_hint=" . urlencode(config('azure.domain_hint')) . "&scope=" . urldecode(config('azure.scope')) . '&redirect_uri=' . urlencode(route('azure.callback'));
```
Comment (danieltjewett) [inline README.md]: Make sure to update to new code to account for the callback.

**Review by danieltjewett — CHANGES_REQUESTED** [review]
@jimmypuckett Thanks for contributing!  Now I understand a bit more what the original issue was asking for.  We've always supplied the callback route in the Azure portal.  However, I can see this use case.

> I got to thinking about this change, & it would be breaking for someone that already had the package installed without the named route

Yes, I would agree.  And, for those like us who are using the callback from the portal as oppose this convention would also have potential issues.  I would assume the callback param passed would take precedence over what is in the portal.

My suggestion would be check to see if the named route exists.  If it does, pass the `redirect_uri` key in.

Obviously this changes a bit of the readme and might make the code a bit gross.  I'll note some ways to remedy this in the code.

**Comment by jimmypuckett** [comment]
One note about this...

> Yes, I would agree. And, for those like us who are using the callback from the portal as oppose this convention would also have potential issues. I would assume the callback param passed would take precedence over what is in the portal.

You still have to set up the callback in the portal, and the callback is passed in will not work unless it is in the portal first.  This just allows you to have a single app configured with multiple callbacks for different environments  (local, staging, production, etc).

**Comment by jimmypuckett** [comment]
@danieltjewett I will try to look over these suggestions/changes later tonight.  It is going to be a very busy week, so I will see what I can do.  Thanks for the feedback.

**Comment by danieltjewett** [comment]
@jimmypuckett Not a problem!  I understand the busyness (which is why I took a couple days for me to review).

**Comment by danieltjewett** [comment]
@jimmypuckett Missed the first comment.  Love the idea of having multiple environments.  We've been creating separate apps for different environments.  This is way better :D

File: src/Azure.php (hunk 1 above)
Comment (jimmypuckett) [reply src/Azure.php]: OK, I have just done this in 5b81293ef77659063c32e8f3780eb3b821a3d17e

File: src/Azure.php (hunk 2 above)
Comment (jimmypuckett) [reply src/Azure.php]: Done in 5b81293ef77659063c32e8f3780eb3b821a3d17e

File: README.md (hunk 3 above)
Comment (jimmypuckett) [reply README.md:11]: Done in 5bfdd2b44e382d77066b7518d9804d0d3d8cef30

File: README.md (hunk 4 above)
Comment (jimmypuckett) [reply README.md:95]: Done in 5bfdd2b44e382d77066b7518d9804d0d3d8cef30

File: README.md (hunk 5 above)
Comment (jimmypuckett) [reply README.md]: Done in 5bfdd2b44e382d77066b7518d9804d0d3d8cef30

**Comment by jimmypuckett** [comment]
@danieltjewett I finally got a break to make the tweaks that you suggested.  Please let me know if there is anything else.

Thanks.

**Comment by jimmypuckett** [comment]
@danieltjewett I know that I got tied up & took a bit to implement your suggestions, but do you think that you could review/merge this.  We have been working on a project that is a few weeks from being live, so this would be super nice to have for the UAT & prod environments without having to set up additional applications in the portal.

Thanks!

**Comment by danieltjewett** [comment]
@jimmypuckett sorry about that.  I am actually no longer a part of the rootinc org, and didn't get this notification via normal email.  I just happened to see this in my github notifications.  @jaredrhodes could you review this?

**Comment by jimmypuckett** [comment]
@danieltjewett thanks for the update.

@jaredrhodes if this is not something that you guys have the time to keep alive, I would be happy to take over the support & maintenance of it along with all of our other Laravel packages.  Just let me know.  Thanks.

**Comment by jaredrhodes** [comment]
Hi @jimmypuckett, `#thanksdan` @danieltjewett. Since I'm fresh on this, correct me wrong...

**Summary**: Handle multiple environments within a single App Registration.
- In a single Azure App Registration, one can register multiple "Redirect URIs" like `"https://myapp.test/auth"`, and `"https://staging.myapp.com/auth"` to handle multiple environments.
-  If no `'azure.callback'` named route exists carry on as previous using the only or last "Redirect URI" registered in Azure App Registration.
- If a route is named `'azure.callback'` will be turned into a a URL and sent off in the auth request. If a match is found in Azure App Registration's "Redirect URIs", a valid login will redirect to that URL.

This looks to be backward compatible where if someone does not register a route with `->name('azure.callback');` we carry on as usual and if they do register a route with `->name('azure.callback');` we use that value?

**Comment by jimmypuckett** [comment]
Yes, that is the key concept of the MR, and yes, it should be backward compatible.

**Comment by danieltjewett** [comment]
@jaredrhodes That is correct. :shipit:

**Review by jaredrhodes — APPROVED** [review]

**Pull request merged**